src/server.h

```diff
@@ -429,6 +434,26 @@ extern int configOOMScoreAdjValuesDefaults[CONFIG_OOM_COUNT];
 #define CLIENT_REPLICATION_DONE (1ULL << 51)         /* Indicate that replication has been done on the client */
 #define CLIENT_AUTHENTICATED (1ULL << 52)            /* Indicate a client has successfully authenticated */
 
+#define CLIENT_REPL_MAIN_CONN                                                                                          \
+    (1ULL << 52) /* Dual connection sync: track a connection                                                           \
+                         which is used for online replication data */
+#define CLIENT_REPL_RDB_CONN                                                                                           \
+    (1ULL << 53) /* Dual connection sync: track a connection                                                           \
+                          which is used for rdb snapshot */
+#define CLIENT_PROTECTED_RDB_CONN                                                                                      \
```

Comment:
I am still not 100% sure we need `CLIENT_PROTECTED_RDB_CONN` but I am OK with punting on this discussion and going with your current proposal of having this new state.

Reply:
Not using this flag might cause some sync sessions to fail due to high client write traffic. While this isn't necessarily bad, we should be aware of the consequences.

Copied from previous thread:
TLDR: The `CLIENT_PROTECTED_RDB_CONN` flag's primary role is to preserve the shared replication buffer data blocks, not to protect the client from eviction.

Expanded explanation:

The `CLIENT_PROTECTED_RDB_CONN` flag is used in a specific replication process between a primary and a replica Redis instance. This process involves:

1. The replica requests a full synchronization using the RDB channel.

2. The primary approves this request and sends:
   - The `RDB_END_OFFSET` (indicating where the RDB transfer ends)
   - The actual RDB data

3. The replica completes loading the RDB data via the RDB connection. In the meanwhile, the replica's main connection hasn't yet established a PSYNC with the primary.

Typically, after this step, the primary would free its shared replication buffer at the RDB_END_OFFSET. However, the `CLIENT_PROTECTED_RDB_CONN` flag prevents this from happening. It's used to stop the primary from releasing the replication data blocks that the replica's RDB client was using, even though that RDB client is no longer active at this stage. This protection ensures that necessary replication data remains available for the subsequent synchronization steps.

Reply:
I think this goes back to your previous point that the PSYNC request can't be sent synchronously - assuming so, I agree with you that this flag is required. The thing that I don't 100% understand is why PSYNC can only be sent asynchronously and this is the discussion that I prefer to have after the merge. Let's keep this flag for this PR.